br/pkg/restore/import_mode_switcher.go

```diff
@@ -200,6 +204,87 @@ func RestorePreWork(
 	return mgr.RemoveSchedulersWithConfig(ctx)
 }
 
+func FineGrainedRestorePreWork(
+	ctx context.Context,
+	mgr *conn.Mgr,
+	switcher *ImportModeSwitcher,
+	tableIDs []int64,
+	isOnline bool,
+	switchToImport bool,
+) (pdutil.UndoFunc, *pdutil.ClusterConfig, error) {
+	if isOnline {
+		return pdutil.Nop, nil, nil
+	}
+
+	if switchToImport {
+		// Switch TiKV cluster to import mode (adjust rocksdb configuration).
+		err := switcher.GoSwitchToImportMode(ctx)
+		if err != nil {
+			return pdutil.Nop, nil, err
+		}
+	}
+
+	// pause config
+	originCfg, err := mgr.GetOriginPDConfig(ctx)
+	if err != nil {
+		return pdutil.Nop, nil, err
+	}
+
+	// pause scheduler
+	keyRange := calSortedKeyRanges(tableIDs)
+	ruleID, resumeScheduler, err := mgr.RemoveSchedulersOnRegion(ctx, keyRange)
+	if err != nil {
+		return pdutil.Nop, nil, err
+	}
+	originCfg.RuleID = ruleID
+
+	// handle undo
+	undo := mgr.MakeFineGrainedUndoFunction(originCfg, resumeScheduler)
+	return undo, &originCfg, errors.Trace(err)
+}
+
+func calSortedKeyRanges(ids []int64) [][2]kv.Key {
+	idRanges := calSortedTableIds(ids)
+	if len(idRanges) == 0 {
+		return [][2]kv.Key{}
+	}
+	keyRanges := [][2]kv.Key{}
+	for i := range idRanges {
+		startKey := tablecodec.EncodeTablePrefix(idRanges[i][0])
+		startKey = codec.EncodeBytes([]byte{}, startKey)
+		endKey := tablecodec.EncodeTablePrefix(idRanges[i][1])
+		endKey = codec.EncodeBytes([]byte{}, endKey)
+		keyRanges = append(keyRanges, [2]kv.Key{startKey, endKey})
+	}
+	return keyRanges
+}
+
+func calSortedTableIds(ids []int64) [][]int64 {
+	if len(ids) == 0 {
+		return [][]int64{}
+	}
+
+	slices.Sort(ids)
+
+	var idRanges [][]int64
+
+	start := ids[0]
+	end := start + 1
+
+	for i := 1; i < len(ids); i++ {
+		if ids[i] <= ids[i-1]+2 {
```

Comment:
I would say a safer/effecient approach would be allocate all the ids at once and assign those ids to tables before creation, that way we only need to set rule for that one range. 
But that involves more code change as right now we allocate only up to max upstream id, to achieve that we need to allocate more for those tables that cannot reuse table id. 
@Leavrth let me know what you think.